### PR TITLE
Fixed two bugs in the bloom_filter index.

### PR DESCRIPTION
**Pull request by ghost**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):

I fixed two bugs in the bloom_filter index.
1. Bloom_filter performs too many redundant index calculations.
2. Bloom_filter can't handle any zero length array.

Detailed description (optional):

## Redundant index calculations
I could improve performance after fixing this bug.

**Create a test table**
```SQL
CREATE TABLE test_bf
(
    `id` int, 
    `ary` Array(String), 
    INDEX idx_ary ary TYPE bloom_filter(0.01) GRANULARITY 8192
)
ENGINE = MergeTree()
ORDER BY id

```

**Insert data**
```SQL
INSERT INTO test_bf SELECT 
    number AS id, 
    [CAST(id, 'String'), CAST(id + 1, 'String'), CAST(id + 2, 'String')]
FROM system.numbers
LIMIT 1000000
```

The insert query takes `19sec` in the current version.
```
0 rows in set. Elapsed: 19.020 sec. Processed 1.05 million rows, 8.39 MB (55.13 thousand rows/s., 441.04 KB/s.)
```

But, after fixed this bug, it takes only `0.611sec`.
```
0 rows in set. Elapsed: 0.611 sec. Processed 1.05 million rows, 8.39 MB (1.71 million rows/s., 13.72 MB/s.)
```

## Zero length array
Current version of bloom_filter threw segmentation fault exception when I inserted a zero length array.

```SQL
INSERT INTO test_bf VALUES (0, [])
-- Server crashes by segmentation fault exception while processing the insert query.
```

So I fixed this bugs.

Thank you.

### CODE REVIEW

File: dbms/src/Interpreters/BloomFilterHash.h [hunk 1]
```diff
@@ -85,8 +85,15 @@ struct BloomFilterHash
                 throw Exception("Unexpected type " + data_type->getName() + " of bloom filter index.", ErrorCodes::LOGICAL_ERROR);
 
             const auto & offsets = array_col->getOffsets();
-            size_t offset = (pos == 0) ? 0 : offsets[pos - 1];
-            limit = std::max(array_col->getData().size() - offset, limit);
+            limit = offsets[pos + limit - 1] - ((pos == 0) ? 0 : offsets[pos - 1]);
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/BloomFilterHash.h]: FYI, we allow to access PaddedPODArray at index -1.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
> I could improve performance after fixing this bug.

Could you please add this into `dbms/tests/performance`?

File: dbms/tests/performance/bloom_filter.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+    <stop_conditions>
+        <all_of>
+            <total_time_ms>30000</total_time_ms>
+        </all_of>
+    </stop_conditions>
+
+    <create_query>DROP TABLE IF EXISTS test_bf</create_query>
+    <create_query>CREATE TABLE test_bf (`id` int, `ary` Array(String), INDEX idx_ary ary TYPE bloom_filter(0.01) GRANULARITY 8192) ENGINE = MergeTree() ORDER BY id</create_query>
+    <query>INSERT INTO test_bf SELECT number AS id, [CAST(id, 'String'), CAST(id + 1, 'String'), CAST(id + 2, 'String')] FROM system.numbers LIMIT 3000000</query>
```
Comment (alexey-milovidov) [inline dbms/tests/performance/bloom_filter.xml]: Ok.

The result could be non-stable due to merges. In that case, we should surround the test with SYSTEM STOP MERGES, SYSTEM START MERGES.

**Comment by alexey-milovidov** [comment]
Performance test is failed due to `allow_experimental_data_skipping_indices`.
I will make it the default and merge this PR.

**Comment by achimbab** [comment]
@alexey-milovidov 
I learned a lot about ClickHouse.
I surrounded the test with `SYSTEM STOP MERGES`, `SYSTEM START MERGES` and set `allow_experimental_data_skipping_indices` for this test.
When you make it the default, I will remove it if needed.
Thank you.

**Comment by achimbab** [comment]
@alexey-milovidov 
It's not an issue for this PR, but how do you think about https://github.com/ClickHouse/ClickHouse/pull/8022?
I made that PR two weeks ago, but nobody left any comment.
I wonder how to do that. :-)

**Comment by alexey-milovidov** [comment]
> It's not an issue for this PR, but how do you think about #8022?
I made that PR two weeks ago, but nobody left any comment.
I wonder how to do that. :-)

It's on queue, we will continue (either me or someone from my team).

**Comment by achimbab** [comment]
@alexey-milovidov 
I fixed additional bug that referencing an invalid offset in bloom_filter with array.
Thank you.

**Comment by achimbab** [comment]
@alexey-milovidov 
Hello, Did you have a merry Christmas?~
I passed all tests and now waiting for your review.
Please let me know If there is any other problem.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Thank you! Everything is perfect.

**Pull request merged**